src/librustc/mir/tcx.rs

```diff
@@ -205,14 +205,19 @@ impl<'tcx> NeoPlace<'tcx> {
         mir: &'cx Mir<'tcx>,
         tcx: &TyCtxt<'cx, 'gcx, 'tcx>,
     ) -> Option<Field> {
-        let (place, by_ref) = if let Some(ProjectionElem::Deref) = self.elems.last() {
-            (self.projection_base(*tcx), true)
-        } else {
-            (self.clone(), false)
-        };
+        let mut by_ref = false;
```

Comment:
Pre-existing: it'd be nice to have a comment explaining what this is matching against. Something like:

---

Look for either `*(P.f)` or `P.f`, where `P` is a place with closure type. These sorts of places represent accesses to the closure's captured upvars.